python/adjoint/Basis.py

```diff
@@ -0,0 +1,440 @@
+## density is the number of basis functions per unit length.
+##
+## (This is a somewhat simplistic implementation that supports only
+## first-order elements over triangles, possibly to be replaced
+## eventually by something more sophisticated from the many available
+## choices of python FEM codes.)
+##
+## How it works: The sides of an (lx x ly) rectangle are divided in to
+## segments at a resolution of `density` functions per unit length.
+## To each node in the resulting (NX+1)*(NY+1) grid of nodes we assign
+## a single basis function. The basis function associated with node #n
+## is only supported on those triangles naving node #n as a vertex;
+## it takes the value 1 at node #n and falls linearly to zero at
+## the neighboring nodes. Indexing: The node with grid coordinates
+## (nx,ny) (0 \le ni \le N_i) is assigned index nx*(NY+1) + ny,
+## following the convention MEEP scheme for grid indexing.
+######################################################################
+class FiniteElementBasis(Basis):
+
+    def __init__(self, lx, ly, density=4):
+        self.l=[lx,ly]
+        self.N=[int(np.ceil(density*ll)) for ll in self.l]
+        self.d=[ll/(1.0*NN) for ll,NN in zip(self.l,self.N)]
+        super().__init__( (self.N[0]+1)*(self.N[1]+1) )
+
+    def in_grid(self, n, ofs=[0,0]):
+        return np.all([nn+oo in range(0,NN+1) for(nn,oo,NN) in zip(n,ofs,self.N)])
+
+    # scalar index of basis function associated with node n + optional offset
+    def bindex(self, n, ofs=[0,0]):
+        return -1 if not self.in_grid(n,ofs) else (n[0]+ofs[0])*(self.N[1]+1) + (n[1]+ofs[1])
+
+    @property
+    def shape(self):
+        return (self.N[0]+1,self.N[1]+1)
+
+    ##############################################################
+    # on input, p[0,1] are the x,y coordinates of an evaluation
+    # point in the grid. The return value is a list of
+    # (bindex,bvalue) pairs giving the index and value of all
+    # basis functions supported at p.
+    ##############################################################
+    def contributors(self, p):
+        pshift       = [ pp + 0.5*ll for pp,ll in zip(p,self.l) ]
+        node         = [ int(np.floor(pp/dd)) for pp,dd in zip(pshift,self.d) ]
+        xi           = [ (pp-nn*dd)/(1.0*dd) for (pp,nn,dd) in zip(pshift,node,self.d) ]
+        xisum, lower = xi[0]+xi[1], xi[0] <= (1.0-xi[1])
+        indices      = [self.bindex(node,ofs) for ofs in [(1,0), (0,1), (0,0) if lower else (1,1)]]
+        vals         = [xi[0], xi[1], 1.0-xisum] if lower else [1.0-xi[1],1.0-xi[0],xisum-1.0]
+        return [ (i,v) for i,v in zip(indices,vals) if i!= -1 ]
+
+    def __call__(self, p=[0.0,0.0]):
+        b=np.zeros(self.dim)
+        for idx,val in self.contributors(p):
+           b[idx]=val
+        return b
+
+    def eval(self, p=[0.0,0.0], beta_vector=None):
+        bv = beta_vector if beta_vector else self.beta_vector
+        return sum( [ bv[idx]*val for idx,val in self.contributors(p) ] )
```

Comment:
Eventually we will want a native C++ implementation of this sort of finite-element basis (or just bilinear interpolation on a grid, rather than triangles), since for optimization you will eventually want to use something like this `eval` for a material function, but a pure-Python implementation like this will be pretty slow when setting up ε on a lot of grid points.   (Especially in practical cases one will want even more complicated basis functions, like the SIMP method with a smoothing filter, in order to do optimization with a "binarization" pass and a resolution constraint.)